### PR TITLE
vkd3d: BLOB PIX markers decoding

### PR DESCRIPTION
**Pull request by Dragon31337**
PIX Events Markers ( https://devblogs.microsoft.com/pix/winpixeventruntime/ ), which is the most used and the only recommended way to insert Perf Markers, see ( https://docs.microsoft.com/en-us/windows/win32/api/d3d12/nf-d3d12-id3d12commandqueue-beginevent )
PIXEventType enum values, masks, offsets, and encoding can be extracted from the source code in the original package
https://www.nuget.org/packages/WinPixEventRuntime
at \Include\WinPixEventRuntime\, mostly in PIXEventsCommon.h

### CODE REVIEW

**Comment by HansKristian-Work** [comment]
We are in a process of relicensing to MIT. Before we can accept this PR, do you consent to your contribution being relicensed to MIT?

**Comment by Dragon31337** [comment]
Yes, I do consent to my contribution being relicensed to MIT.

**Review by HansKristian-Work — CHANGES_REQUESTED** [review]
The formatting here is all over the place.

- Inverse christmas tree for variables.
- Put all variables on top (C89 style).
- 4 space indents.
- Snake case.
- Pointer asterisk ties right.

Also, missing Signed-off-by line (`git commit --amend -s`)

File: libs/vkd3d/command.c [hunk 1]
```diff
@@ -9988,6 +9988,76 @@ static void STDMETHODCALLTYPE d3d12_command_list_SetPredication(d3d12_command_li
...
+  data64++; //skip Color
+  data64++;
+
+  isANSI = *data64 & PIXEventsStringIsANSIReadMask;
+  
+  data64++;
+
+  if (isANSI)
+  {
+    StrLength = strnlen_s((const char*)data64, SIZE_MAX);
```
Comment (Joshua-Ashton) [inline libs/vkd3d/command.c:10044]: Just use `strlen`. It's not useful to use `strnlen` here to protect against iterating over the whole address space. :frog:


FTR: The `_s` functions are originally meme MSVC extensions that got erroneously promoted in C11, when they should have just used the ubiquitous POSIX standard ones. If you do need to use `strnlen` again, just use `strnlen`.

**Review by Joshua-Ashton — CHANGES_REQUESTED** [review]

**Comment by Dragon31337** [comment]
I'm not sure if I should keep PIX consts and enum names or convert those too. I would keep it for reference to the original code in PIXEventsCommon.h

File: libs/vkd3d/command.c [hunk 2]
```diff
@@ -9988,6 +9988,70 @@ static void STDMETHODCALLTYPE d3d12_command_list_SetPredication(d3d12_command_li
...
 }
 
+// *PIXEvent* enum values and consts are from PIXEventsCommon.h of winpixeventruntime package
+
+static char* decode_pix_blob(const void* data, size_t size)
+{
+    static const UINT64 PIXEventsStringIsANSIReadMask = 0x0040000000000000;
+    static const UINT64 PIXEventsTypeReadMask = 0x00000000000FFC00;
+    static const UINT64 PIXEventsTypeBitShift = 10;
+    UINT64* data_uint64_aligned = (UINT64*)(data);
```
Comment (HansKristian-Work) [inline libs/vkd3d/command.c:9998]: `const UINT64 *data_uint64_aligned = data;`

File: libs/vkd3d/command.c [hunk 3]
```diff
@@ -9988,6 +9988,70 @@ static void STDMETHODCALLTYPE d3d12_command_list_SetPredication(d3d12_command_li
     }
 }
 
+// *PIXEvent* enum values and consts are from PIXEventsCommon.h of winpixeventruntime package
+
+static char* decode_pix_blob(const void* data, size_t size)
```
Comment (HansKristian-Work) [inline libs/vkd3d/command.c:9993]: `static char *` and `const void *data`

File: libs/vkd3d/command.c [hunk 4]
```diff
@@ -9988,6 +9988,70 @@ static void STDMETHODCALLTYPE d3d12_command_list_SetPredication(d3d12_command_li
...
+    static const UINT64 PIXEventsTypeBitShift = 10;
+    UINT64* data_uint64_aligned = (UINT64*)(data);
+    size_t label_str_length;
+    char *label_str = NULL;
+    bool is_ansi;
+    UINT64 type;
+
+    enum PIXEventType
+    {
+        ePIXEvent_EndEvent = 0x000,
```
Comment (HansKristian-Work) [inline libs/vkd3d/command.c:10006]: I guess it's fine to keep the enums like this.

File: libs/vkd3d/command.c [hunk 5]
```diff
@@ -9988,6 +9988,70 @@ static void STDMETHODCALLTYPE d3d12_command_list_SetPredication(d3d12_command_li
...
+    if (type == ePIXEvent_SetMarker_NoArgs)
+        type = ePIXEvent_BeginEvent_NoArgs;
+
+    if (type != ePIXEvent_BeginEvent_NoArgs)
+    {
+        WARN("Unexpected/unsupported PIX3Event type, SetMarker_VarArgs and SetMarker_VarArgs can be added on request");
+        return NULL;
+    }
+
+    data_uint64_aligned++; //skip Color
```
Comment (HansKristian-Work) [inline libs/vkd3d/command.c:10030]: Use C89 style comments `/* */`

File: libs/vkd3d/command.c [hunk 6]
```diff
@@ -9988,6 +9988,70 @@ static void STDMETHODCALLTYPE d3d12_command_list_SetPredication(d3d12_command_li
...
+
+    data_uint64_aligned++;
+    
+    if (is_ansi)
+    {
+        label_str_length = strlen((const char*)data_uint64_aligned);
+        label_str = vkd3d_malloc(label_str_length + 1);
+        if (!label_str)
+            return NULL;
+        memcpy(label_str, (const char*)data_uint64_aligned, label_str_length);
```
Comment (HansKristian-Work) [inline libs/vkd3d/command.c:10043]: No need to cast `data_uint64_aligned` here.

File: libs/vkd3d/command.c [hunk 7]
```diff
@@ -9988,6 +9988,70 @@ static void STDMETHODCALLTYPE d3d12_command_list_SetPredication(d3d12_command_li
...
+        label_str[label_str_length] = '\0';
+    }
+    else
+    {
+        label_str_length = wcslen((const wchar_t*)data_uint64_aligned);
```
Comment (HansKristian-Work) [inline libs/vkd3d/command.c:10048]: Don't bother checking length here.

File: libs/vkd3d/command.c [hunk 8]
```diff
@@ -9988,6 +9988,70 @@ static void STDMETHODCALLTYPE d3d12_command_list_SetPredication(d3d12_command_li
...
+        label_str = vkd3d_strdup_w_utf8((const wchar_t*)data_uint64_aligned, label_str_length);
```
Comment (HansKristian-Work) [inline libs/vkd3d/command.c:10049]: Deduce the length based on `size`, subtracted by the blob offset. Also it's `WCHAR` not `wchar_t`. The C type is 32-bit on Linux, but we explicitly need 16-bit.

**Review by HansKristian-Work — CHANGES_REQUESTED** [review]

File: libs/vkd3d/command.c (hunk 7 above)
Comment (DadSchoorse) [inline libs/vkd3d/command.c:10048]: Use `vkd3d_wcslen((const WCHAR*) ...)` WCHAR is always utf16. wchar_t is platform dependent.

Comment (DadSchoorse) [reply libs/vkd3d/command.c:10048]: actually, you can drop this completely. vkd3d_strdup_w_utf8 will stop at a null terminator.

File: libs/vkd3d/command.c (hunk 8 above)
Comment (DadSchoorse) [inline libs/vkd3d/command.c:10049]: `vkd3d_strdup_w_utf8((const WCHAR *)data_uint64_aligned, 0)`

File: libs/vkd3d/command.c [hunk 9]
```diff
@@ -9988,6 +9988,70 @@ static void STDMETHODCALLTYPE d3d12_command_list_SetPredication(d3d12_command_li
...
+    data_uint64_aligned++;
+    
+    if (is_ansi)
+    {
+        label_str_length = strlen((const char*)data_uint64_aligned);
+        label_str = vkd3d_malloc(label_str_length + 1);
+        if (!label_str)
+            return NULL;
+        memcpy(label_str, (const char*)data_uint64_aligned, label_str_length);
+        label_str[label_str_length] = '\0';
```
Comment (DadSchoorse) [inline libs/vkd3d/command.c:10044]: Just use `vkd3d_strdup`.

File: libs/vkd3d/command.c (hunk 7 above)
Comment (DadSchoorse) [reply libs/vkd3d/command.c:10048]: actually I ignore me and use the blob length like themaister said

**Comment by Dragon31337** [comment]
THanks for your help and feedback. Are there any items left?

**Review by HansKristian-Work — APPROVED** [review]
LGTM

**Pull request merged**